python/thunder/utils/decorators.py

```diff
@@ -0,0 +1,281 @@
+
+            Arguments
+
+              numpy-storage: choose one of ['auto', 'ascii', 'base64'] (default: auto)
+
+              Use the 'numpyStorage' argument to select whether numpy arrays
+              will be encoded in ASCII (as a list of lists) in Base64 (i.e.
+              space efficient binary), or to select automatically (the default)
+              depending on the size of the array. Currently the Base64 encoding
+              is selecting if the array has more than 1000 elements.
+
+            Returns
+
+              The object encoded as a python dictionary with "JSON-safe" datatypes that is ready to
+              be converted to a string using Python's standard JSON library (or another library of
+              your choice).
+            '''
+            from collections import OrderedDict
+            from numpy import ndarray
+
+            def serializeRecursively(data):
+                import datetime
+
+                if data is None or type(data) == bool or type(data) == int or \
+                   type(data) == long or type(data) == float or type(data) == str:
+                    return data
+                if type(data) == list:
+                    return [serializeRecursively(val) for val in data]           # Recurse into lists
+                if type(data) == OrderedDict:
+                    return {"py/collections.OrderedDict":
+                            [[serializeRecursively(k), serializeRecursively(v)] for k, v in data.iteritems()]}
+                if _isNamedTuple(data):
+                    return {"py/collections.namedtuple": {
+                        "type":   type(data).__name__,
+                        "fields": list(data._fields),
+                        "values": [serializeRecursively(getattr(data, f)) for f in data._fields]}}
+                if type(data) == dict:
+                    if all(type(k) == str for k in data):   # Recurse into dicts
+                        return {k: serializeRecursively(v) for k, v in data.iteritems()}
+                    else:
+                        return {"py/dict": [[serializeRecursively(k), serializeRecursively(v)] for k, v in data.iteritems()]}
+                if type(data) == tuple:                          # Recurse into tuples
+                    return {"py/tuple": [serializeRecursively(val) for val in data]}
+                if type(data) == set:                            # Recurse into sets
+                    return {"py/set": [serializeRecursively(val) for val in data]}
+                if type(data) == datetime.datetime:
+                    return {"py/datetime": str(data)}
+                if type(data) == complex:
+                    return {"py/complex": [ data.real, data.imag] }
+                if type(data) == ndarray:
+                    if numpyStorage == 'ascii' or (numpyStorage == 'auto' and data.size < 1000):
+                        return {"py/numpy.ndarray.ascii": {
```

Comment:
Any concern that this "type" is of our own invention and not a real python or numpy type, unlike say `collections.namedtuple`? I don't have a better suggestion and think I'm fine with it, just throwing it out there.